typing/includecore.ml

```diff
@@ -429,16 +426,22 @@ let type_declarations ?(equality = false) ~loc env ~mark name
           mark usage cstrs1;
           if equality then mark Env.Positive cstrs2
         end;
-        Option.map
-          (fun var_err -> Variant_mismatch var_err)
-          (compare_variants ~loc env decl1.type_params decl2.type_params 1
-             cstrs1 cstrs2)
+        begin match rep1, rep2 with
+        | Variant_unboxed, Variant_regular ->
+           Some (Unboxed_representation First)
+        | Variant_regular, Variant_unboxed ->
+           Some (Unboxed_representation Second)
+        | _ ->
+          Option.map
+            (fun var_err -> Variant_mismatch var_err)
+            (compare_variants ~loc env decl1.type_params decl2.type_params 1
+               cstrs1 cstrs2)
+        end
```

Comment:
I think that we want a `compare_variants_with_representation` help here, by symmetry with the record case.

Reply:
Done. I ended up rewriting both of these functions.